src/components/TodoApp.tsx

```diff
@@ -0,0 +1,223 @@
+import { TEMP_TODO_ID } from '../constants/tempTodoId';
+
+export const TodoApp = () => {
+  const {
+    todos,
+    errorMessage,
+    tempTodo,
+  } = useContext(StateContext);
+  const dispatch = useContext(DispatchContext);
+
+  const inputRef = useRef<HTMLInputElement | null>(null);
+  const completedTodosCount = todos.filter(
+    todo => todo.completed,
+  ).length;
+
+  const [title, setTitle] = useState('');
+  const [isFocus, setIsFocus] = useState(false);
+  let isCompletedAll = completedTodosCount === todos.length;
+
+  const toggleAll = () => {
+    isCompletedAll = !isCompletedAll;
+
+    dispatch({
+      type: 'shouldLoading',
+      payload: LoadingStatus.All,
+    });
+
+    const todosForToggle = isCompletedAll
+      ? todos.filter(t => !t.completed)
+      : todos.filter(t => t.completed);
+
+    todosForToggle.forEach(async todo => {
+      try {
+        let updatedTodo = todo;
+
+        updatedTodo = {
+          ...todo,
+          completed: isCompletedAll,
+        };
+        await updateTodo(updatedTodo);
+
+        dispatch({
+          type: 'updateTodo',
+          payload: {
+            todo: updatedTodo,
+            loadingType: LoadingStatus.None,
+          },
+        });
+      } catch (error) {
+        isCompletedAll = !isCompletedAll;
+
+        dispatch({
+          type: 'error',
+          payload: {
+            error: ErrorMessage.Updating,
+            loadingType: LoadingStatus.None,
+          },
+        });
+      }
+    });
+  };
+
+  const createNewTodo = async (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    if (!title.trim()) {
+      dispatch({
+        type: 'error',
+        payload: { error: ErrorMessage.EmptyTitle },
+      });
+      setIsFocus(true);
+
+      return;
+    }
+
+    try {
+      const newTodo = {
+        title: title.trim(),
+        userId: USER_ID,
+        completed: false,
+      };
+
+      dispatch({
+        type: 'createTempTodo',
+        payload: {
+          todo: { ...newTodo, id: TEMP_TODO_ID },
+          loadingType: LoadingStatus.Current,
+        },
+      });
+
+      const responseNewTodo = await createTodo(newTodo);
+
+      dispatch({
+        type: 'createTodo',
+        payload: {
+          todo: responseNewTodo,
+          loadingType: LoadingStatus.None,
+        },
+      });
+
+      setTitle('');
+    } catch (error) {
+      dispatch({
+        type: 'error',
+        payload: {
+          error: ErrorMessage.Creating,
+          loadingType: LoadingStatus.None,
+        },
+      });
+    } finally {
+      dispatch({
+        type: 'createTempTodo',
+        payload: {
+          todo: null,
+          loadingType: LoadingStatus.None,
+        },
+      });
+      setIsFocus(true);
+    }
+  };
+
+  const closeErrorMessage = () => {
+    dispatch({ type: 'error', payload: { error: ErrorMessage.None } });
+  };
+
+  useEffect(() => {
+    if (errorMessage) {
+      setTimeout(() => dispatch(
```

Comment:
when you use setTimeout or setInterval at useEffect you should clear this operations with `return () => {....}`